src/modem.rs

```diff
@@ -0,0 +1,335 @@
+use at_commands::{builder::CommandBuilder, parser::CommandParser};
+use atat::{self, AtatCmd};
+use core::{borrow::BorrowMut, ffi::c_void, marker::PhantomData};
+use esp_idf_hal::{
+    delay::{TickType, BLOCK},
+    io::EspIOError,
+    uart::{UartDriver, UartTxDriver},
+};
+
+use crate::{
+    eventloop::EspSystemEventLoop,
+    handle::RawHandle,
+    netif::{EspNetif, EspNetifDriver, NetifStack},
+    sys::*,
+};
+
+pub struct EspModem<'d, T>
+where
+    T: BorrowMut<UartDriver<'d>>,
+{
+    serial: T,
+
+    _d: PhantomData<&'d ()>,
+}
+
+impl<'d, T> EspModem<'d, T>
+where
+    T: BorrowMut<UartDriver<'d>> + Send,
+{
+    pub fn new(serial: T) -> Self {
+        Self {
+            serial,
+            _d: PhantomData,
+        }
+    }
+
+    pub fn send_cmd<CMD: AtatCmd>(&mut self, cmd: &CMD) -> Result<CMD::Response, atat::Error> {
+        let mut buff = [0u8; 64];
+        // flush the channel
+        // self.serial
+        //     .borrow_mut()
+        //     .clear_rx()
+        //     .map_err(|_err| atat::Error::Write)?;
+
+        // write the command to the uart
+        let len = cmd.write(&mut buff);
+        log::info!("about to write {:?}", &buff[..len]);
+        self.serial
+            .borrow_mut()
+            .write(&buff[..len])
+            .map_err(|_err| atat::Error::Write)?;
+
+        // now read the uart to get the response
+
+        let len = self
+            .serial
+            .borrow_mut()
+            .read(&mut buff, TickType::new_millis(1000).ticks())
+            .map_err(|_err| atat::Error::Read)?;
+        log::info!("got response {:?}", &buff[..len]);
+        cmd.parse(Ok(&buff[..len]))
+    }
+
+    pub fn setup_data_mode(&mut self, sysloop: EspSystemEventLoop) -> Result<(), EspError> {
```

Comment:
Given that this function does _more_ than just setting up data mode - i.e. it "runs" the PPP stack RX/TX loop, I suggest to rename it to simply "run". It should also run "forever" or until the PPP TX/RX loop exits because it had lost connection to the modem, or because the loop receives a PPP frame that designates the modem had exited.

Reply:
... or - if we decide that `EspModem/SimModem` should only be platform-neutral, "pure Rust" and only dealing with the "pre-PPP" phase, we can leave it named `setup_data_mode` OR maybe `negotiate`.

Reply:
I have created a `run` loop, but the only issue is that it mutably borrows. I need to work around that so that one can call "is_connected" etc

Reply:
@ivmarkov any ideas for the mutable borrow here. I want to run the `run` loop in a separate thread (spawned in userland), but I also want to do things like `EspModem.wait_netif_up()` after the `run` loop, but this requires another borrow. You can see my failed attempts in the `examples/lte_modem.rs`

Reply:
Using an `Arc<Mutex<...>>` **externally** to `EspModem::run` will be impossible, because the mutex will lock the modem for as long as the `run` method "runs" so to say, which is not what we want.

So if we have any interior mutability (i.e. `Mutex` or similar) it has to be internal to the modem struct itself.

I think we have two options:
- Option A: (I have a slight prefernce for this) Turn `run(&mut self)` into `run(&self)` and then internally use `Mutex`es (i.e. interior mutability)
- Option B: Introduce `EspModem::split(&mut self) -> (EspModemRunner<'_>, EspModemStatus<'_>), where `EspModemRunner` just has a `EspModemRunner::run(&mut self)` and `EspModemStatus` has whatever you need in terms of status (`is_connected` etc.)

Given that both options will likely require some form of interior mutability after all (i.e. a `Mutex` or suchlike) I would say perhaps Option A?

Reply:
I like option A, although it seems like we would need to internally Arc+mutex the `netif` and `serial` members